fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -82,30 +82,49 @@ Future<Void> forwardValue(Promise<T> out, Future<T> in)
 
 int getBytes(Promise<Version> const& r) { return 0; }
 
-ACTOR Future<Void> getRate(UID myID, MasterInterface master, int64_t* inTransactionCount, double* outTransactionRate) {
-	state Future<Void> nextRequestTimer = Void();
+ACTOR Future<Void> getRate(UID myID, MasterInterface master, int64_t* inTransactionCount, double* outTransactionRate, HealthMetrics* healthMetrics,
+						   GetDetailedHealthMetricsReply* getDetailedHealthMetricsReply) {
+	state Future<Void> nextRequestTimer = Never();
+	state Future<Void> nextDetailedRequestTimer = Void();
 	state Future<Void> leaseTimeout = Never();
+	state Future<Void> detailedLeaseTimeout = Never();
 	state Future<GetRateInfoReply> reply;
 	state int64_t lastTC = 0;
 
 	loop choose{
 		when(wait(nextRequestTimer)) {
 			nextRequestTimer = Never();
-			reply = brokenPromiseToNever(master.getRateInfo.getReply(GetRateInfoRequest(myID, *inTransactionCount)));
+			reply = brokenPromiseToNever(master.getRateInfo.getReply(GetRateInfoRequest(myID, *inTransactionCount, false)));
+		}
+		when(wait(nextDetailedRequestTimer)) {
+			nextDetailedRequestTimer = Never();
+			nextRequestTimer = Never();
+			reply = brokenPromiseToNever(master.getRateInfo.getReply(GetRateInfoRequest(myID, *inTransactionCount, true)));
 		}
 		when(GetRateInfoReply rep = wait(reply)) {
 			reply = Never();
+			healthMetrics->update(rep.healthMetrics, rep.detailed, true);
 			*outTransactionRate = rep.transactionRate;
 			//TraceEvent("MasterProxyRate", myID).detail("Rate", rep.transactionRate).detail("Lease", rep.leaseDuration).detail("ReleasedTransactions", *inTransactionCount - lastTC);
 			lastTC = *inTransactionCount;
 			leaseTimeout = delay(rep.leaseDuration);
 			nextRequestTimer = delayJittered(rep.leaseDuration / 2);
+			if (rep.detailed) {
+				detailedLeaseTimeout = delay(SERVER_KNOBS->DETAILED_METRIC_UPDATE_RATE);
+				nextDetailedRequestTimer = delayJittered(SERVER_KNOBS->DETAILED_METRIC_UPDATE_RATE / 2);
+				getDetailedHealthMetricsReply->setHealthMetrics(*healthMetrics);
+			}
 		}
 		when(wait(leaseTimeout)) {
 			*outTransactionRate = 0;
 			//TraceEvent("MasterProxyRate", myID).detail("Rate", 0).detail("Lease", "Expired");
 			leaseTimeout = Never();
 		}
+		when(wait(detailedLeaseTimeout)) {
```

Comment:
The way this is written, if we do not get a detailed reply within 5 seconds, the proxy will stop giving out GRVs to clients. I do not think this is necessary. A better way to modify this loop for detailed metrics would be to just keep a double for the last time we have gotten a detailed reply, and if more than DETAILED_METRIC_UPDATE_RATE - rep.leaseDuration time has elapsed since the reply we should ask for a detailed reply instead of normal one.